storage/innobase/handler/ha_innodb.cc

```diff
@@ -5153,10 +5153,40 @@ UNIV_INTERN void lock_cancel_waiting_and_release(lock_t* lock);
 
 /** Cancel any pending lock request associated with the current THD.
 @sa THD::awake() @sa ha_kill_query() */
-static void innobase_kill_query(handlerton*, THD* thd, enum thd_kill_levels)
+static void innobase_kill_query(handlerton*, THD* thd, enum thd_kill_levels level)
 {
 	DBUG_ENTER("innobase_kill_query");
 #ifdef WITH_WSREP
+	if (level == THD_WSREP_MARK_VICTIM)
+	{
+		if (trx_t *trx= thd_to_trx(thd))
+		{
+			/*
+			  not a direct kill, but just marking the victim for
+			  later aborting locking mutexes in agreed locking
+			  order, last to lock is victim thd LOCK_thd_data
+			  mutex victim thd's wsrep_aborter is set while holding
+			  all mutexes. We miss to see if victim was already
+			  marked for abort by some other killer but that does
+			  not matter here.
+			  Finally releasing all mutexes in reverse order
+			*/
+			lock_mutex_enter();
+			trx_sys_mutex_enter();
+			trx_mutex_enter(trx);
+			wsrep_thd_LOCK(thd);
+			if (wsrep_thd_set_wsrep_aborter(thd, thd))
+			{
+				WSREP_DEBUG("innodb kill transaction skipped");
+			}
+			wsrep_thd_UNLOCK(thd);
+			trx_mutex_exit(trx);
+			trx_sys_mutex_exit();
+			lock_mutex_exit();
+			DBUG_VOID_RETURN;
+		}
+		else DBUG_VOID_RETURN;
```

Comment:
We should follow same mutexing as when real kill happens, i.e. trx_sys.mutex is not needed.

Reply:
Please do not mark any review comments as ‘resolved’ until they actually are resolved. As far as I can tell, neither of your promises have been addressed by revising the pull request. And you did not answer my questions regarding the implementation of the `wsrep_`  functions. We only have stubs for those functions in the MariaDB server code base.

Reply:
Naturally, as I have not yet committed any revisions. Implementation of wsrepthd_LOCK() is sql/wsrep_mysqld.cc and wsrep_thd_set_wsrep_aborter() is on sql/wsrep_thd.cc. We need those InnoDB mutexes to avoid concurrent brute force lock conflict resolution. I will do 10.3 and 10.4 versions when I'm sure that this implementation is correct. I'm not yet there.